jax/_src/util.py

```diff
@@ -285,8 +286,8 @@ def merge(new_lhs, new_rhs):
 
   return lhs, rhs, merge
 
-def cache(max_size=4096):
-  def wrap(f):
+def cache(max_size: int | None = 4096) -> Callable[[F], F]:
```

Comment:
Note that this signature doesn't allow calling e.g. `cache_clear` on the resulting callable.

Reply:
Good point... we could probably fix that with `ParamSpec`, though I worry we'd run into the same issues that are blocking #14688

Reply:
I don't think `ParamSpec`s would help here. The feature we need is called intersection types, and unfortunately it's still in the design phase. If we did have intersection types, we could say that

```
def cache(...) -> Callable[[F], F&CacheLike]: ...
```

where

```
class CacheLike(Protocol):
  def cache_clear(self) -> None: ...
  # ...
```

I'd say let's submit this unless mypy/pytype complain :)

Reply:
pytype complains about the lack of `cache_clear` and `cache_info` methods. I managed to fix that using ParamSpec, but then either pytype breaks (if I try to mark the output type as covariant) or mypy breaks (if I don't mark the output type as covariant).

It seems there's no way to annotate this in a way that's compatible with both `mypy` and `pytype`, so I think I'm going to abandon the effort.

Reply:
SGTM.

Reply:
Well, you *can* write mypy specific code.

```
MYPY = False  # mypy secretly set this to true

if MYPY:
 ...
else:
  ...
```